src/Confluent.Kafka.Avro/GenericDeserializerImpl.cs

```diff
@@ -56,30 +58,29 @@ public GenericRecord Deserialize(string topic, byte[] array)
                 }
                 var writerId = IPAddress.NetworkToHostOrder(reader.ReadInt32());
 
-                // Note: there is a race condition here in which more than one datumReader can
-                // be created for a given writerId, but this is benign (doesn't matter) and 
-                // happens very infrequently => doesn't affect the effectiveness of the cache.
-
-                datumReaderBySchemaId.TryGetValue(writerId, out DatumReader<GenericRecord> datumReader);
-                if (datumReader == null)
+                lock (deserializeLockObj)
                 {
-                    // TODO: If any of this cache fills up, this is probably an
-                    // indication of misuse of the deserializer. Ideally we would do 
-                    // something more sophisticated than the below + not allow 
-                    // the misuse to keep happening without warning.
-                    if (datumReaderBySchemaId.Count > schemaRegistryClient.MaxCachedSchemas)
+                    datumReaderBySchemaId.TryGetValue(writerId, out DatumReader<GenericRecord> datumReader);
+                    if (datumReader == null)
                     {
-                        datumReaderBySchemaId.Clear();
-                    }
+                        // TODO: If any of this cache fills up, this is probably an
+                        // indication of misuse of the deserializer. Ideally we would do 
+                        // something more sophisticated than the below + not allow 
+                        // the misuse to keep happening without warning.
+                        if (datumReaderBySchemaId.Count > schemaRegistryClient.MaxCachedSchemas)
+                        {
+                            datumReaderBySchemaId.Clear();
+                        }
 
-                    var writerSchemaJson = schemaRegistryClient.GetSchemaAsync(writerId).Result;
-                    var writerSchema = Avro.Schema.Parse(writerSchemaJson);
+                        var writerSchemaJson = schemaRegistryClient.GetSchemaAsync(writerId).ConfigureAwait(false).GetAwaiter().GetResult();
+                        var writerSchema = Avro.Schema.Parse(writerSchemaJson);
 
-                    datumReader = new GenericReader<GenericRecord>(writerSchema, writerSchema);
-                    datumReaderBySchemaId[writerId] = datumReader;
-                }
+                        datumReader = new GenericReader<GenericRecord>(writerSchema, writerSchema);
+                        datumReaderBySchemaId[writerId] = datumReader;
+                    }
 
-                return datumReader.Read(default(GenericRecord), new BinaryDecoder(stream));
+                    return datumReader.Read(default(GenericRecord), new BinaryDecoder(stream));
```

Comment:
shouldnt this be outside the lock?

Reply:
you're right, it's better outside the lock, though doesn't need to be for correctness.